src/reformat/alpha.rs

```diff
@@ -158,44 +183,29 @@ impl image::Image {
         if self.planes2[3].is_none() {
             return Ok(());
         }
-        if !self.planes2[3].as_ref().unwrap().is_pointer() {
-            // TODO: implement this function for non-pointer inputs.
-            return Err(AvifError::NotImplemented);
-        }
-        let src = image::Image {
-            width: self.width,
-            height: self.height,
-            depth: self.depth,
-            yuv_format: self.yuv_format,
-            planes2: [
-                None,
-                None,
-                None,
-                Some(Pixels::Pointer(self.planes2[3].as_ref().unwrap().pointer())),
-            ],
-            row_bytes: [0, 0, 0, self.row_bytes[3]],
-            ..image::Image::default()
-        };
-        self.allocate_planes(Category::Alpha)?;
         let width = self.width as usize;
         let depth = self.depth;
-        if depth > 8 {
-            for y in 0..self.height {
-                let src_row = src.row16(Plane::A, y)?;
-                let dst_row = self.row16_mut(Plane::A, y)?;
-                for x in 0..width {
-                    dst_row[x] = limited_to_full_y(depth, src_row[x]);
+        match depth {
+            10 | 12 => {
+                for y in 0..self.height {
+                    let row = self.row16_mut(Plane::A, y)?;
+                    for x in 0..width {
+                        row[x] = limited_to_full_y(depth, row[x]);
+                    }
                 }
             }
-        } else {
-            for y in 0..self.height {
-                let src_row = src.row(Plane::A, y)?;
-                let dst_row = self.row_mut(Plane::A, y)?;
-                for x in 0..width {
-                    dst_row[x] = limited_to_full_y(8, src_row[x] as u16) as u8;
+            8 => {
+                for y in 0..self.height {
+                    let row = self.row_mut(Plane::A, y)?;
+                    for x in 0..width {
+                        row[x] = limited_to_full_y(8, row[x] as u16) as u8;
```

Comment:
we cannot reformat alpha in place if the buffer type is a pointer.

in that case the buffers are stolen from the underlying decoder. so it will corrupt the decoder's reference buffers (animation or progressive or any other case where the decoder needs to use the current frame as a reference frame).

the TODO about implementing this for non-pointer inputs can be fixed this way however (i.e.) the new code can be added as a separate case where the input type is not Pointer.

you can probably revert this change and deal with this in a separate PR.

Reply:
I see there is no test for that issue.

Could this be enforced at compile time?  
Maybe three kinds of `Pixels`:
- mutable `Pointer`
- read-only `Pointer`
- `Buffer` or `Buffer16`

Reply:
Also I did not expect `alpha_to_full_range()` to create a buffer. In my opinion it should modify samples in-place if possible or return an error. The calling site should make a copy first if needed. What do you think?

Reply:
> I see there is no test for that issue.
> 
> Could this be enforced at compile time? Maybe three kinds of `Pixels`:
> 
> * mutable `Pointer`
> * read-only `Pointer`
> * `Buffer` or `Buffer16`

I don't think there is a need to distinguish between mutable and read-only pointers. Data is stored as a pointer only when it comes from the underlying C/C++ library. So in that case we should always treat it as immutable. Any time we need to modify it, we would to convert it into either Buffer or Buffer16.

It is similar to `imageOwnfYUVPlanes` flag in the `avifImage` struct in libavif. In case of pointer, the flag can be considered to be false and in case of Buffer or Buffer16, the flag can be considered to be true.

> Also I did not expect alpha_to_full_range() to create a buffer. In my opinion it should modify samples in-place if possible or return an error. The calling site should make a copy first if needed. What do you think?

The alpha_to_full_range is a bit of an old case and only exists for providing backwards compatibility. The latest version of the AVIF spec mandates that alpha plane always has to be in full range. This function exists to support files that may have been mistakenly created with older versions of libavif which allowed setting alpha range to limited. There is only one calling site, so it doesn't really matter if we do the allocation here or in the calling site. So maybe we can just add a comment that says this function will re-allocate the buffers if necessary.

Reply:
> Any time we need to modify it, we would to convert it into either Buffer or Buffer16.

What about decoding a grid? The convenient way is usually making each tile a view to the reconstructed image buffer. I can see mutable `Pointer` as a solution for that. But I guess slices to a `Buffer` would be an implementation closer to Rust paradigms and would be safer. However I can imagine `image::Image` becoming too complex because of the lifetime tracking. What do you think?

> It is similar to `imageOwnfYUVPlanes` flag in the `avifImage` struct in libavif. In case of pointer, the flag can be considered to be false and in case of Buffer or Buffer16, the flag can be considered to be true.

The C language used for libavif does not provide the same memory security guarantees as Rust. In CrabbyAvif I was a bit disappointed to learn that I could just modify the function, half guessing I fixed it somehow, mess with external memory without touching any unsafe code and still compile and make tests pass. That is why I would prefer an `image::Image` implementation that does not compile if one tries to modify pixels pointed to by a `Pointer`. It would prevent this mistake in the future.  
Can it be done technically? `row_mut()` could just fail for a `Pointer` no?

> So maybe we can just add a comment that says this function will re-allocate the buffers if necessary.

Sounds good. Thank you for the explanation.

Reply:
> > Any time we need to modify it, we would to convert it into either Buffer or Buffer16.
> 
> What about decoding a grid? The convenient way is usually making each tile a view to the reconstructed image buffer. I can see mutable `Pointer` as a solution for that. But I guess slices to a `Buffer` would be an implementation closer to Rust paradigms and would be safer. However I can imagine `image::Image` becoming too complex because of the lifetime tracking. What do you think?
> 

For a grid image, we will have to make a copy of the plane buffers and cannot use the underlying decoder's buffers as a pointer. The main reason for this is that the underlying buffer will have a stride that is different from the width of the tile and we cannot support per-tile stride. So we will have to copy to ensure we have exactly `width` pixels in each tile.

For grid images, the plane buffers will already be of type Buffer or Buffer16 and they are mutable. So we should just be able to do this conversion in place (the allocate_planes call will not re-allocate the buffers in this case). I think that is how the current implementation works.

> > It is similar to `imageOwnfYUVPlanes` flag in the `avifImage` struct in libavif. In case of pointer, the flag can be considered to be false and in case of Buffer or Buffer16, the flag can be considered to be true.
> 
> The C language used for libavif does not provide the same memory security guarantees as Rust. In CrabbyAvif I was a bit disappointed to learn that I could just modify the function, half guessing I fixed it somehow, mess with external memory without touching any unsafe code and still compile and make tests pass. That is why I would prefer an `image::Image` implementation that does not compile if one tries to modify pixels pointed to by a `Pointer`. It would prevent this mistake in the future. Can it be done technically? `row_mut()` could just fail for a `Pointer` no?
> 

Yes, making `row_mut` fail for pointers is a good idea. But there might be places where we need to pass a mutable pointer because the C/C++ function signatures don't have them marked as `const`.

> > So maybe we can just add a comment that says this function will re-allocate the buffers if necessary.
> 
> Sounds good. Thank you for the explanation.